README.md

```diff
@@ -1,111 +1,282 @@
-        decls.NewIdent("you", decls.String, nil))
-    env, err := cel.NewEnv(decls)
-    if err != nil {
-        log.Fatalf("environment creation error: %s\n", err)
-    }
-
-    // Parse and type-check the expression.
-    //
-    // Detailed information about which issues/errors were encountered may be
-    // found within the `iss`, but a non-nil `iss.Err()` value will contain a
-    // detailed error string in human-readable form.
-    p, iss := env.Parse(`"Hello " + you + "! I'm " + i + "."`)
-    if iss != nil && iss.Err() != nil {
-        log.Fatalln(iss.Err())
-    }
-    c, iss := env.Check(p)
-    if iss != nil && iss.Err() != nil {
-        log.Fatalln(iss.Err())
-    }
-
-    // Create the program, and evaluate it against some input.
-    prg, err := env.Program(c)
-    if err != nil {
-        log.Fatalf("program creation error: %s\n", err)
-    }
-
-    // Note, the second result, the EvalDetails will be non-nil when using
-    // either cel.OptTrackState or cel.OptExhaustiveEval (not shown). The
-    // details value holds information about the evaluation, rather than the
-    // output of the evaluation itself.
-    out, _, err := prg.Eval(cel.Vars(map[string]interface{}{
-        "i": "CEL",
-        "you": "world"}))
-    if err != nil {
-        log.Fatalf("runtime error: %s\n", err)
-    }
-
-    // Hello world! I'm CEL.
-    fmt.Println(out)
+env, err := cel.NewEnv(
+    cel.Declarations(
+        decls.NewIdent("name", decls.String, nil),
+        decls.NewIdent("group", decls.String, nil))
+```
+
+That's it, the environment is ready to be use for parsing and type-checking.
+CEL supports all the usual primitive types in addition to lists, maps, as well
+as first-class support for JSON and Protocol Buffers.
+
+### Parse and Check
+
+The parsing phase indicates whether the expression is syntactically valid and
+expands any macros present within the environment. Parsing and checking is
+more computationally expensive than evaluation, and it is recommended that
+expressions be parsed and checked ahead of time.
+
+```go
+parsed, issues := env.Parse(`name.startsWith("/groups/" + group)`)
+if issues != nil && issues.Err() != nil {
+    log.Fatalf("parse error: %s", issues.Err())
+}
+checked, issues := env.Check(parsed)
+if issues != nil && issues.Err() != nil {
+    log.Fatalf("type-check error: %s", issues.Err())
 }
+prg, err := env.Program(checked)
+if err != nil {
+    log.Fatalf("program construction error: %s", err)
+}
+```
+
+The `cel.Program` generated at the end of parse and check is stateless,
+thread-safe, and cachable.
+
+Type-checking in an optional, but strongly encouraged step that can reject some
+semantically invalid expressions using static analysis. Additionally, the check
+produces metadata which can improve function invocation performance and object
+field selection at evaluation-time.
+
+#### Macros
+
+Macros are enabled by default and may be disabled. The comprehension macros
```

Comment:
I've always found macros to be one of the more confusing aspects of CEL. Why are they different than just any other language feature? Why are they called that? I confess that I tried to see how I'd explain and wasn't sure I could.